### PR TITLE
Add Windows build

### PR DESCRIPTION
**Pull request by donny-dont**
This adds a `pipeline.libsonnet` file which is included by both `.drone.jsonnet` and `.drone.windows.jsonnet` to create the associated yaml files.

### CODE REVIEW

File: .drone.windows.jsonnet [hunk 1]
```diff
@@ -0,0 +1,9 @@
+local pipeline = import 'pipeline.libsonnet';
+local name = 'drone-webhook';
+
+[
+  pipeline.test('windows', 'amd64', '1803'),
```
Comment (tboerger) [inline .drone.windows.jsonnet]: IMHO we could skip test and vet on windows

File: .drone.jsonnet [hunk 2]
```diff
@@ -1,169 +1,14 @@
...
+local name = 'drone-webhook';
 
 [
-  PipelineTesting,
-  PipelineBuild("linux", "amd64"),
-  PipelineBuild("linux", "arm64"),
-  PipelineBuild("linux", "arm"),
-  PipelineNotifications,
+  pipeline.test('linux', 'amd64'),
+  //pipeline.build(name, 'linux', 'amd64'),
```
Comment (tboerger) [inline .drone.jsonnet]: Why have this been commented out?

File: pipeline.libsonnet [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+        {
+          name: 'vet',
+          image: golang,
+          pull: 'always',
+          environment: {
+            GO111MODULE: 'on',
+          },
+          commands: [
+            'go vet ./...',
+          ],
```
Comment (tboerger) [inline pipeline.libsonnet]: We should add a volume:

```
          volumes: [
            {
              name: 'gopath',
              path: '/go',
            }
          ],
```

File: pipeline.libsonnet [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+        {
+          name: 'test',
+          image: golang,
+          pull: 'always',
+          environment: {
+            GO111MODULE: 'on',
+          },
+          commands: [
+            'go test -cover ./...',
+          ],
```
Comment (tboerger) [inline pipeline.libsonnet]: Same here for the volume.

File: pipeline.libsonnet [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+        },
+      ],
```
Comment (tboerger) [inline pipeline.libsonnet]: Here we need the volume definition:

```
      volumes: [
        {
          name: 'gopath',
          temp: {}
        },
      ],
```

File: pipeline.libsonnet [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+    },
+
+  build(name, os='linux', arch='amd64', version='')::
+    local is_windows = windows(os);
+    local tag = if is_windows then os + '-' + version else os + '-' + arch;
+    local file_suffix = std.strReplace(tag, '-', '.');
+    local volumes = if is_windows then [{ name: windows_pipe_volume, path: windows_pipe }] else [];
+    local golang = golang_image(os, version);
+    local plugin_repo = 'plugins/' + std.split(name, '-')[1];
```
Comment (tboerger) [inline pipeline.libsonnet]: We could just supply the name `webhook`, then we don't need to split it. Besides that, this logic will break for plugins like `drone-volume-cache`.

File: pipeline.libsonnet [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+          when: {
+            event: ['tag'],
+          },
+        },
+        {
+          name: 'executable',
+          image: golang,
+          pull: 'always',
+          commands: [
+            './release/' + os + '/' + arch + '/' + name + extension + ' --help',
```
Comment (tboerger) [inline pipeline.libsonnet]: Does this also work on Windows? Or do we need to rewrite it for Windows?

File: pipeline.libsonnet [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+            auto_tag_suffix: tag,
+            daemon_off: if is_windows then 'true' else 'false',
+            dockerfile: 'docker/Dockerfile.' + file_suffix,
+            repo: plugin_repo,
+            username: { from_secret: 'docker_username' },
+            password: { from_secret: 'docker_password' },
+          },
+          volumes: if std.length(volumes) > 0 then volumes,
+          when: {
+            event: ['push'],
```
Comment (tboerger) [inline pipeline.libsonnet]:
```patch
-           event: ['push'],
+           event: ['push',  'tag'],
```

File: pipeline.libsonnet [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+          },
+        },
+      ],
+      trigger: {
```
Comment (tboerger) [inline pipeline.libsonnet]: Currently this won't trigger builds for tags, but @bradrydzewski wants to add `ref` as a trigger, than we could do something like this:

```
      trigger: {
        ref: ['refs/heads/master', 'refs/pulls/**', 'refs/tags/**'],
      },
```

File: pipeline.libsonnet [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+      name: 'notifications',
+      platform: {
+        os: os,
+        arch: arch,
+        version: if std.length(version) > 0 then version,
+      },
+      steps: [
+        {
+          name: 'manifest',
+          image: 'plugins/manifest:1',
```
Comment (tboerger) [inline pipeline.libsonnet]: `:latest`? Or keeping `:1`?

File: pipeline.libsonnet [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+            spec: 'docker/manifest.tmpl',
+            ignore_missing: true,
+          },
+          when: {
+            event: ['push'],
+          },
+        },
+        {
+          name: 'microbadger',
+          image: 'plugins/webhook:1',
```
Comment (tboerger) [inline pipeline.libsonnet]: `:latest`? Or keeping `:1`?

File: pipeline.libsonnet [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+          name: 'manifest',
+          image: 'plugins/manifest:1',
+          pull: 'always',
+          settings: {
+            username: { from_secret: 'docker_username' },
+            password: { from_secret: 'docker_password' },
+            spec: 'docker/manifest.tmpl',
+            ignore_missing: true,
+          },
+          when: {
```
Comment (tboerger) [inline pipeline.libsonnet]: Drop this `when` block.

File: pipeline.libsonnet [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+          },
+        },
+        {
+          name: 'microbadger',
+          image: 'plugins/webhook:1',
+          pull: 'always',
+          settings: {
+            url: { from_secret: 'microbadger_url' },
+          },
+          when: {
```
Comment (tboerger) [inline pipeline.libsonnet]: Drop this `when` block.

File: pipeline.libsonnet [hunk 14]
```diff
@@ -0,0 +1,187 @@
...
+            event: ['push'],
+          },
+        },
+      ],
+      depends_on: depends_on,
+      trigger: {
```
Comment (tboerger) [inline pipeline.libsonnet]: Same as above, we could do something like this:

```
      trigger: {
        ref: ['refs/heads/master', 'refs/pulls/**', 'refs/tags/**'],
      },
```

When we are dropping the `when` block from above we should also add this:

```
      trigger: {
        event: ['push', 'tag'],
      },
```

File: pipeline.libsonnet (hunk 7 above)
Comment (donny-dont) [reply pipeline.libsonnet]: Yea it works

File: pipeline.libsonnet (hunk 6 above)
Comment (donny-dont) [reply pipeline.libsonnet]: std.splitLimit(str, c, maxsplits) does the trick.

File: .drone.windows.jsonnet (hunk 1 above)
Comment (donny-dont) [reply .drone.windows.jsonnet]: Eh its not a big deal to test the thing on Windows as well. There's always the chance for some platform specific tests.

File: .drone.jsonnet (hunk 2 above)
Comment (donny-dont) [reply .drone.jsonnet]: There was an issue for a bit with some of the libsonnet files so I was commenting things out and mistakenly forgot to uncomment them.

**Comment by donny-dont** [comment]
@tboerger this is ready to look at.

File: .drone.windows.jsonnet [hunk 15]
```diff
@@ -0,0 +1,9 @@
...
+  pipeline.build(name, 'windows', 'amd64', '1803'),
+  pipeline.build(name, 'windows', 'amd64', '1809'),
+  pipeline.notifications('windows', 'amd64', '1809', ['windows-1803', 'windows-1809']),
```
Comment (tboerger) [inline .drone.windows.jsonnet]: Tests running on 1803, this on 1809, is that intentional?

File: .drone.windows.yml [hunk 16]
```diff
@@ -1,30 +1,57 @@
...
   arch: amd64
+  version: 1803
+
+steps:
+- name: vet
+  pull: always
+  image: golang:1.11-windowsservercore-1803
+  commands:
+  - go vet ./...
+  environment:
```
Comment (tboerger) [inline .drone.windows.yml]: Still missing a volume?

File: .drone.windows.yml [hunk 17]
```diff
@@ -1,30 +1,57 @@
...
+    GO111MODULE: on
+
+- name: test
+  pull: always
+  image: golang:1.11-windowsservercore-1803
+  commands:
+  - go test -cover ./...
+  environment:
```
Comment (tboerger) [inline .drone.windows.yml]: Still missing a volume?

File: .drone.windows.yml (hunk 16 above)
Comment (donny-dont) [reply .drone.windows.yml]: Added logic for windows volume with golang

**Comment by donny-dont** [comment]
K think its good to go @tboerger but you're going to need to merge it yourself.

The PR build is failing due to the `trigger` not having `ref` yet in cloud.drone.io. @bradrydzewski will need to work that out.

**Comment by donny-dont** [comment]
@tboerger brad updated cloud so refs work. Redid the builds. I'm not authorized to merge.

**Pull request merged**